File: javascript/algorithms/heapsort.js
const heapSort = (arr) => {
	const n = arr.length;

	for (let i = (n/2 - 1); i >= 0; i--) 
		heapify(arr, n, i);

	for (let i = n-1; i >= 0; i--) {
		let swap = arr[0];
		arr[0] = arr[i];
		arr[i] = swap;
		heapify(arr, i, 0)
	}

	return arr;
}

const heapify = (arr, n, i) => {
	let largest = i, l = 2*i + 1, r = 2*i + 2;

	if(l < n && arr[l] > arr[largest]) 
		largest = l;

	if(r < n && arr[r] > arr[largest]) 
		largest = r;

	if (largest != i) {
		let swap = arr[i];
		arr[i] = arr[largest];
		arr[largest] = swap;
		heapify(arr, n, largest);
	}
}

File: python/algorithms/heap_Sort.py
def heapsorting(arr, n, i):
   largest = i # largest value
   l = 2 * i + 1 # left
   r = 2 * i + 2 # right

   if l < n and arr[i] < arr[l]:
      largest = l

   if r < n and arr[largest] < arr[r]:
      largest = r

   if largest != i:
      arr[i],arr[largest] = arr[largest],arr[i] # swap
      # root.
      heapsorting(arr, n, largest)

def heapSort(arr):
   n = len(arr)
   
   for i in range(n, -1, -1):
      heapsorting(arr, n, i)

   for i in range(n-1, 0, -1):
      arr[i], arr[0] = arr[0], arr[i] # swap
      heapsorting(arr, i, 0)

arr = [2,5,3,8,6,5,4,7]
heapSort(arr)
n = len(arr)
print ("Sorted array is")
for i in range(n):
   print (arr[i],end=" ")
